# Aim:
## To find the best combination for number of Hidden layers and Activation Functions using
###  Linear and 
### Non-Linear Functions

In [1]:
import numpy as np
import pandas as pd 
import tensorflow as tf
from tensorflow import keras
import matplotlib.pyplot as plt 

In [2]:
(x_train,y_train),(x_test,y_test) = keras.datasets.mnist.load_data()
print("size: ",x_train.shape)

#Scaling the obtained values for better accuracy
x_train = x_train/255
x_test = x_test/255
#converting 2D array into an 1D array using pandas function
x_train_flattened = x_train.reshape(len(x_train), 28*28) 
x_test_flattened = x_test.reshape(len(x_test), 28*28)
print("shapes: ",x_train_flattened.shape, x_test_flattened.shape)


size:  (60000, 28, 28)
shapes:  (60000, 784) (10000, 784)


In [3]:
def myLinear(x):
    return x

## Single Layer model with Linear function

In [4]:
#Sequentila means stack of layers
#Dense means that the every neuron in the layer is connected to every neuron in the forwawrd layer
simple_model = keras.Sequential([
    keras.layers.Dense(10, input_shape=(784,), activation=myLinear)
])

#Compiling the neural network model using Adaptive movement Estimation algorithm; adam
simple_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

#training the model 5 times using the training data set
simple_model.fit(x_train_flattened,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 6s 2ms/step - loss: 2.2457 - accuracy: 0.3946
Epoch 2/5
1875/1875 [==============================] - 5s 2ms/step - loss: 2.1605 - accuracy: 0.3744
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.1749 - accuracy: 0.3386
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.1850 - accuracy: 0.3348
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.1641 - accuracy: 0.3375


## Two Layered Network with all Linear activation Functiuons

In [5]:
#Dense means that the every neuron in the layer is connected to every neuron in the forwawrd layer
simple_model = keras.Sequential([
    keras.layers.Dense(100,input_shape=(784,),activation=myLinear),
    keras.layers.Dense(10, activation=myLinear)
])

#Compiling the neural network model using Adaptive movement Estimation algorithm; adam
simple_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

#training the model 5 times using the training data set
simple_model.fit(x_train_flattened,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 2.1863 - accuracy: 0.5574
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.1914 - accuracy: 0.5036
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.2106 - accuracy: 0.4254
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.2019 - accuracy: 0.2751
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.1814 - accuracy: 0.2872


## Network with Non-Linear hidden layer and Linear Output layer

In [6]:
simple_model = keras.Sequential([
    keras.layers.Dense(100,input_shape=(784,),activation='sigmoid'),
    keras.layers.Dense(10, activation=myLinear)
])

#Compiling the neural network model using Adaptive movement Estimation algorithm; adam
simple_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

#training the model 5 times using the training data set
simple_model.fit(x_train_flattened,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3274 - accuracy: 0.2558
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.2656 - accuracy: 0.2163
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 2.2108 - accuracy: 0.1575
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.2146 - accuracy: 0.1425
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.2343 - accuracy: 0.1329


# Three Layered Network 
## Non-Linear : Linear : Linear(OUTPUT)

In [7]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation='sigmoid'),
    keras.layers.Dense(184,activation=myLinear),
    keras.layers.Dense(10, activation=myLinear)
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 2.3147 - accuracy: 0.0984
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.0981
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.0981
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.0981
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.0981


## Linear : Non-Linear : Linear(OUTPUT)

In [8]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation=myLinear),
    keras.layers.Dense(184,activation='sigmoid'),
    keras.layers.Dense(10, activation=myLinear)
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 2.3290 - accuracy: 0.1435
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.1464
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3026 - accuracy: 0.1464
Epoch 4/5
1875/1875 [==============================] - 5s 2ms/step - loss: 2.3026 - accuracy: 0.1464
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 2.3026 - accuracy: 0.1464


## Linear : Linear : Non-Linear(OUTPUT)

In [9]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation=myLinear),
    keras.layers.Dense(184,activation=myLinear),
    keras.layers.Dense(10, activation='sigmoid')
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3951 - accuracy: 0.8864
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3339 - accuracy: 0.9060
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3175 - accuracy: 0.9111
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3052 - accuracy: 0.9127
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3002 - accuracy: 0.9154


## Non-Linear : Linear : Non-Linear(OUTPUT)

In [10]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation='sigmoid'),
    keras.layers.Dense(184,activation=myLinear),
    keras.layers.Dense(10, activation='sigmoid')
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3271 - accuracy: 0.9022
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1408 - accuracy: 0.9577
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0906 - accuracy: 0.9725
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0638 - accuracy: 0.9790
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0482 - accuracy: 0.9841


## Linear : Non-Linear : Non-Linear(OUTPUT)

In [11]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation=myLinear),
    keras.layers.Dense(184,activation='sigmoid'),
    keras.layers.Dense(10, activation='sigmoid')
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2855 - accuracy: 0.9177
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1550 - accuracy: 0.9533
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1200 - accuracy: 0.9642
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1016 - accuracy: 0.9696
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0889 - accuracy: 0.9720


## Non-Linear : Non-Linear : Linear(OUTPUT)

In [12]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation='sigmoid'),
    keras.layers.Dense(184,activation='sigmoid'),
    keras.layers.Dense(10, activation=myLinear)
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 6.5822 - accuracy: 0.1185
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 6.6821 - accuracy: 0.1186
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 6.8422 - accuracy: 0.1740
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 6.8278 - accuracy: 0.1780
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 6.7887 - accuracy: 0.1129


## All Non-Linear Layers

In [13]:
#adding hidden layer to improve perforance
hidden_layered_model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)), #for not having to use flattened array
    keras.layers.Dense(364,activation='sigmoid'),
    keras.layers.Dense(184,activation='sigmoid'),
    keras.layers.Dense(10, activation='sigmoid')
])

hidden_layered_model.compile(optimizer='adam', 
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

hidden_layered_model.fit(x_train,y_train,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3553 - accuracy: 0.8979
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1483 - accuracy: 0.9557
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0972 - accuracy: 0.9704
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0685 - accuracy: 0.9788
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0507 - accuracy: 0.9842


# Conclusion
## The best Combinations for Layers and Activation Function are :
1. Non-Linear hiddden layer + Linear Hidden layer + Non-Linear Output layer: accuracy = 98.42%
2. All Non-Linear Layer: accuracy = 92.42%  